R/drop_download.R

```diff
@@ -0,0 +1,128 @@
+#' Download a file from Dropbox to disk.
+#'
+#' @param path path to a file in Dropbox
+#' @param local_path path to save file to. If NULL (the default), saves file to working directory with same name. If not null, but a valid folder, file will be saved in this folder with same basename as path. If not null and not a folder, file will be saved to this path exactly.
+#' @param overwrite If TRUE, overwrite local file. Defaults to FALSE
+#' @param progress If TRUE, show a progress bar for large file downloads. Defaults to TRUE in interactive sessions, otherwise FALSE.
+#' @param verbose if TRUE, emit message giving location and size of the newly downloaded file. Defaults to TRUE in interactive sessions, otherwise FALSE.
+#' @template token
+#'
+#' @return path to location file downloaded to, invisibly
+#'
+#' @examples \dontrun{
+#'
+#'   # download a file to the current working directory
+#'   drop_get("dataset.zip")
+#'
+#'   # download again, overwriting previous result
+#'   drop_get("dataset.zip", overwrite = TRUE)
+#'
+#'   # download to a different path, keeping file name constant
+#'   # will download to "some/other/place/dataset.zip"
+#'   drop_get("dataset.zip", local_file = "some/other/place/")
+#'
+#'   # download to to a different path, changing filename
+#'   drop_get("dataset.zip", local_file = "some/other/place/not_a_dataset.zip")
+#' }
+#'
+#' @export
+drop_download <- function(
+  path,
+  local_path = NULL,
+  overwrite = FALSE,
+  progress = interactive(),
+  verbose = interactive(),
+  dtoken = get_dropbox_token()
+) {
+
+  # if path isn't an id or revision, ensure it has leading slash
+  if (!grepl("^(id|rev):", path)) path <- add_slashes(path)
+
+  # if no local path given, download it to working directory
+  # if path given is folder, append filename to it
+  if (is.null(local_path)) {
+    local_path = basename(path)
+  } else if (dir.exists(local_path)) {
+    local_path <- file.path(local_path, basename(path))
+  }
+
+  url <- "https://content.dropboxapi.com/2/files/download"
+
+  req <- httr::POST(
+    url = url,
+    httr::config(token = dtoken),
+    httr::add_headers("Dropbox-API-Arg" = jsonlite::toJSON(
+      list(
+        path = path
+      ),
+      auto_unbox = TRUE
+    )),
+    httr::write_disk(local_path, overwrite)
+  )
+
+  httr::stop_for_status(req)
+
+  # print message in verbose mode
+  if (verbose) {
+
+    size <- file.size(local_path)
+    class(size) <- "object_size"
+
+    message(sprintf(
+      "Downloaded %s to %s: %s on disk",
+      path,
+      local_path,
+      format(size, units = "auto")
+    ))
+  }
+
+  # return full path to file, invisibly
+  invisible(as.character(req$content))
```

Comment:
This also returns
```
 Error: $ operator is invalid for atomic vectors 
```
which is why I would also suggest returning

```
invisible(req$content)
```

Reply:
I'm confused, what's the suggestion here? leave it or change it?

If "change it", I would use `message` over `cat` (never use `cat` in a package, user can't suppress it), and I'd be wary of using `message` without a `verbose = TRUE` type option because of how that pollutes the non-interactive experience, i.e. shiny logs

Reply:
plus I already print a message if `verbose`

Reply:
ok. I understand about the logs. But right now `invisible(as.character(req$content))` returns an error for me (even though the download actually occurs).  So I'm asking that you replace it with:

```
  # return full path to file, invisibly
  invisible(req$content)
```
![image](https://user-images.githubusercontent.com/138494/30831674-2702f1c0-a1fd-11e7-8562-b9d2a48a404a.png)

Reply:
hmm...this will break all the tests, as the `as.character` is what I used to coerce it away from being a  special "path" object. Seems especially weird for this to vary among platforms. Does it happen only if `progress = TRUE`?

Reply:
This is resolved as some weird interaction in my `.rprofile`. So please ignore.